File: ocaml/kmeans.ml
open Core.Std
open Point

let min_by f xs =
  let Some h = List.hd xs in
  let m = f h in
  let (_, min_elt) = List.fold xs ~init:(m, h)
    ~f:(fun (a, x) y -> let b = f y in (if b < a then (b, y) else (a, x))) in
  min_elt

let closest p qs = min_by (dist p) qs

let sum qs = List.fold qs ~init:{x = 0.0; y = 0.0} ~f:( ++ )

let average qs = (sum qs) // (qs |> List.length |> Float.of_int)

let group_by xs f =
  let table = Hashtbl.Poly.create () in
  List.iter xs (fun x ->
    let y = f x in
    match Hashtbl.find table y with
      | Some l -> Hashtbl.replace table ~key:y ~data:(x :: l)
      | None -> Hashtbl.replace table ~key:y ~data:[x]);
  table

let clusters xs centroids =
  group_by xs (fun p -> closest p centroids)
  |> Hashtbl.data

let run points iters n =
  let centroids = ref (List.take points n) in
  let new_clusters = ref (clusters points !centroids) in
  for i = 1 to iters do
    centroids := List.map !new_clusters average;
    new_clusters := clusters points !centroids;
  done;
  !new_clusters

File: haskell/Kmeans.hs
module Kmeans where
import Data.List (minimumBy)
import Data.Ord (comparing)
import Data.Map (Map, fromListWith, elems)
import Debug.Trace (trace)
import Point

closest :: Point -> [Point] -> Point
closest p qs = minimumBy (comparing (dist p)) qs

groupBy :: Ord b => (a -> b) -> [a] -> Map b [a]
groupBy f xs = fromListWith (++) [(f x, [x]) | x <- xs]

clusters :: [Point] -> [Point] -> [[Point]]
clusters ps centroids = elems $ groupBy (flip closest centroids) ps

run :: [Point] -> Int -> Int -> [[Point]]
run ps iters n = step ps centroids iters
  where
    centroids = take n ps
    step ps centroids iters =
      let newClusters = clusters ps centroids in
      if iters == 0 then newClusters -- trace (show centroids) newClusters
      else step ps (map average newClusters) (iters - 1)